Power management stat : add verbosity level for MBED_SLEEP_TRACING_ENABLED

jeromecoutant opened the pull request:
### Summary of changes <!-- Required -->

Using MBED_SLEEP_TRACING_ENABLED is not reliable with STM32 targets.

See discussions in #11497 #14580 ...

[edit]
~~This new MBED_SLEEP_STAT_ENABLED macro is the same one~~
~~removing prints at each sleep_manager_lock/unlock_deep_sleep~~

- deepsleep stats can be enabled now with json config
- default configuration is not changed (full verbosity adding a console line for each lock/unlock command)
- except for STM32 targets, verbosity is reduced by default

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

STM32 readme file is updated

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

@ARMmbed/team-st-mcd 

----------------------------------------------------------------------------------------------------------------

0xc0170 commented on platform/include/platform/mbed_power_mgmt.h:
```diff
@@ -90,7 +90,7 @@ extern "C" {
  */
 
-#ifdef MBED_SLEEP_TRACING_ENABLED
+#if defined(MBED_SLEEP_TRACING_ENABLED) || defined(MBED_SLEEP_STAT_ENABLED)
```
I need to read the discussion but still this one commit does not explain one why there would be a need for additional macro (there is || between so one of them needs t obe truth at least,why ? ).

jeromecoutant replied:
Because ST LPTICKER API calls the deepsleep lock function very very often:
https://github.com/ARMmbed/mbed-os/blob/master/targets/TARGET_STM/lp_ticker.c#L471

So I let the current macro as it is  and as it is described in mbed-os doc,
and created a new one, a bit less verbose.

0xc0170 commented on targets/TARGET_STM/README.md:
```diff
@@ -377,6 +409,33 @@ You can change this in you local mbed_app.json:
+- deepsleep can also be disabled by application or drivers using sleep_manager_lock_deep_sleep()
+
+To enable the tracing, it is recommended to **not** define MBED_SLEEP_TRACING_ENABLED macro,
+but **MBED_SLEEP_STAT_ENABLED** macro.
```
To my question above, why is not recommended/ Shoudn't the one be fixed rather than create two (it might confuse people) ?

0xc0170 replied:
Add here some details why stats only.

0xc0170 commented on platform/source/mbed_power_mgmt.c:
```diff
@@ -239,6 +243,9 @@ void sleep_manager_sleep_auto(void)
 #endif
         hal_deepsleep();
     } else {
+#if defined(MBED_SLEEP_STAT_ENABLED)
```
you get stats at the top of this function, do we need to print  this again when only sleeping?

0xc0170 reviewed (commented):
`MBED_SLEEP_STATS_ENABLED` could it be named similarly as the functions in the power mgmt

Not against introducing the stats only. My understanding is tracing in this case for stm32 is way to verbose (happens very frequently) and thus can't be used. Stats however are useful information and can be printed. What a user should do if they need tracing (are there any options for them) ?

Please review @ARMmbed/mbed-os-core review

jeromecoutant committed:
Power management stat : add verbosity level for MBED_SLEEP_TRACING_ENABLED

Full verbosity is adding a console line for each lock/unlock API call

- stats can be enabled with json config
- default configuration is full verbosity and add a console line for each lock/unlock command
- for STM32 targets, verbosity is reduced by default

LDong-Arm reviewed (commented):
I personally prefer this PR to #15046. Locking and unlocking of sleep/deepsleep are fairly common in target-specific drivers/HAL implementations. With reduced verbosity (i.e. the macro not enabled), we only get logs from `sleep_tracker_add` and `sleep_tracker_print_stats` but not every lock/unlock, which is still a reasonable use case.

I think it's better to have both `MBED_SLEEP_TRACING_ENABLED` and `MBED_SLEEP_STAT_ENABLED` configurable and documented inside platform/mbed_lib.json.

jeromecoutant committed:
STM32: readme update for MBED_SLEEP_TRACING_ENABLED

0xc0170 commented on platform/source/mbed_power_mgmt.c:
```diff
@@ -138,19 +141,28 @@ static sleep_statistic_t *sleep_tracker_add(const char *const filename)
+            }
+
+            if (sleep_stats[i].identifier[0] == '\0') {
+                mbed_error_printf("allo");
```
Why is it `allo` ? Should this be something more meaningful ?

0xc0170 replied:
I reviewed this but kept the review pending (not certain why it was not posted).

jeromecoutant replied:
oups...

0xc0170 reviewed (dismissed):
The printf there with allo does not provide sufficient meaning to a reader. What should be printed?

LDong-Arm reviewed (approved):
LGTM as per previous discussion regarding the documentation of the new config.
